### PR TITLE
PCSCHEMA *pc_schema_from_xml(const char *)

### PR DESCRIPTION
**Pull request by mbredif**
This PR is based on a slight rework of #108 from @achidlow to streamline ressource freeing and provide uniform return values in `pc_schema_from_xml`.
This PR goes further by returning the schema pointer (with NULL on failure) instead of the PC_FAILURE/SUCESS enum, to be provide a nicer api. 

I have just rebased the PR for review but accidentally squashed the commits of @achidlow. If anyone is willing to review it and merge it, maybe its history should be rewritten to give proper credit to @achidlow.

### CODE REVIEW

File: lib/pc_schema.c [hunk 1]
```diff
@@ -594,13 +561,21 @@ pc_schema_from_xml(const char *xml_str, PCSCHEMA **schema)
...
-	xmlFreeDoc(xml_doc);
+cleanup:
+	if(s && !schema)
+		pc_schema_free(s);
+	if(xpath_obj)
+		xmlXPathFreeObject(xpath_obj);
+	if(xpath_ctx)
+		xmlXPathFreeContext(xpath_ctx);
+	if(xml_doc)
+		xmlFreeDoc(xml_doc);
```
Comment (elemoine) [inline lib/pc_schema.c]: Styling: spaces are missing.

**Comment by elemoine** [comment]
I think this deserves mulitple commits.

- one for the `pc_schema_from_xml` signature change
- one for the new error handling in `pc_schema_from_xml` and related tests

**Comment by mbredif** [comment]
done !

File: lib/cunit/cu_tester.c [hunk 2]
```diff
@@ -72,7 +72,7 @@ int main(int argc, char *argv[])
 	/* Set up to use the system memory management / logging */
 	pc_install_default_handlers();
 
-	pc_set_handlers(0, 0, 0, cu_error_reporter, 0, 0);
+	pc_set_handlers(0, 0, 0, cu_error_reporter, 0, cu_error_reporter);
```
Comment (elemoine) [inline lib/cunit/cu_tester.c]: What does that do?

File: lib/pc_schema.c [hunk 3]
```diff
@@ -367,18 +367,6 @@ void pc_schema_check_xyzm(PCSCHEMA *s)
...
-	{
-		pcerror("pc_schema_check_xyzm: invalid x_position '%d'", s->x_position);
-		return;
-	}
-
-	if ( s->y_position < 0 )
-	{
-		pcerror("pc_schema_check_xyzm: invalid y_position '%d'", s->y_position);
-		return;
-	}
```
Comment (elemoine) [inline lib/pc_schema.c]: What's the reason for removing this?

File: lib/pc_schema.c [hunk 4]
```diff
@@ -594,13 +562,21 @@ pc_schema_from_xml(const char *xml_str, PCSCHEMA **schema)
 		}
 	}
 
-	xmlXPathFreeObject(xpath_obj);
+	if( pc_schema_is_valid(s) )
```
Comment (elemoine) [inline lib/pc_schema.c]: Space after `if`. (The tools should enforce our code styling but this is not the case currently.)

File: lib/pc_schema.c [hunk 5]
```diff
@@ -594,13 +562,21 @@ pc_schema_from_xml(const char *xml_str, PCSCHEMA **schema)
...
+		*schema = s;
 
-	xmlXPathFreeContext(xpath_ctx);
-	xmlFreeDoc(xml_doc);
+cleanup:
+	if( s && !*schema )
```
Comment (elemoine) [inline lib/pc_schema.c]: Space after `if`.

File: lib/pc_schema.c [hunk 6]
```diff
@@ -594,13 +562,21 @@ pc_schema_from_xml(const char *xml_str, PCSCHEMA **schema)
...
+		pc_schema_free(s);
+	if( xpath_obj )
```
Comment (elemoine) [inline lib/pc_schema.c]: Ditto.

File: lib/pc_schema.c [hunk 7]
```diff
@@ -594,13 +562,21 @@ pc_schema_from_xml(const char *xml_str, PCSCHEMA **schema)
...
+		xmlXPathFreeObject(xpath_obj);
+	if( xpath_ctx )
+		xmlXPathFreeContext(xpath_ctx);
+	if( xml_doc )
```
Comment (elemoine) [inline lib/pc_schema.c]: Ditto.

File: lib/pc_schema.c [hunk 8]
```diff
@@ -594,13 +562,21 @@ pc_schema_from_xml(const char *xml_str, PCSCHEMA **schema)
...
+		*schema = s;
 
-	xmlXPathFreeContext(xpath_ctx);
-	xmlFreeDoc(xml_doc);
+cleanup:
+	if( s && !*schema )
+		pc_schema_free(s);
+	if( xpath_obj )
+		xmlXPathFreeObject(xpath_obj);
+	if( xpath_ctx )
```
Comment (elemoine) [inline lib/pc_schema.c]: Ditto.

File: lib/cunit/cu_tester.c (hunk 2 above)
Comment (mbredif) [inline lib/cunit/cu_tester.c]: That was a change by @achidlow that sets the warn_handler. He may have needed it when he was trying to do his python wrapping. I will remove that change.

File: lib/pc_schema.c (hunk 3 above)
Comment (mbredif) [inline lib/pc_schema.c]: It is redundant with [pc_schema_is_valid](https://github.com/mbredif/pointcloud/blob/4ad548561cc0de82bee4229e7d8693b0e9df7cee/lib/pc_schema.c#L587-L597), which is now [enforced](https://github.com/mbredif/pointcloud/blob/4ad548561cc0de82bee4229e7d8693b0e9df7cee/lib/cunit/cu_pc_schema.c#L161-L173) by pc_schema_from_xml.

This was actually the main trouble for @achidlow, as the unhappy path was heterogeneous : pcerror failure here and PC_FAILURE return value in pc_schema_from_xml.

File: lib/pc_schema.c (hunk 5 above)
Comment (mbredif) [inline lib/pc_schema.c]: Ok, is the coding style documented somewhere ?

Comment (elemoine) [inline lib/pc_schema.c]: Not that I know, but this is what's used in most places.

File: lib/cunit/cu_tester.c (hunk 2 above)
Comment (mbredif) [inline lib/cunit/cu_tester.c]: done

File: lib/pc_schema.c (hunk 4 above)
Comment (mbredif) [inline lib/pc_schema.c]: done

Comment (elemoine) [inline lib/pc_schema.c]: It looks that particular one is not done.

**Comment by mbredif** [comment]
Done. You are so picky :)

**Comment by elemoine** [comment]
> Done. You are so picky :)

Probably too much. We need tools to enforce styling.

**Comment by mbredif** [comment]
@achidlow, @strk, @pblottiere, any objection ?

**Comment by pblottiere** [comment]
It seems good to me. I prefer the api this way!

Thank you @achidlow and @mbredif!

**Comment by strk** [comment]
I trust your judgement

**Comment by elemoine** [comment]
One small concern: this patch changes the signature of `pc_schema_from_xml` which is part of the pointcloud library API. Do you think that's a problem? Are there users of the pointcloud C lib?

@mbredif @pblottiere ^^

**Comment by mbredif** [comment]
> Are there users of the pointcloud C lib?

Not that I know. @Remi-C and @pramsey?

**Comment by pramsey** [comment]
Dear god I hope not. It's meant for internal use *only*, to make running
cunit tests easier.

**Pull request merged**